### PR TITLE
Tighten exclusion of floating point mask to enable most common mask types (those with 0 and -inf only)

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Tighten exclusion of floating point mask to enable most common mask types for inference fastpath (those with 0 and -inf only)

Test Plan: sandcastle, github

Differential Revision: D49135963

### CODE REVIEW

**Review by mikaylagawarecki — COMMENTED** [review]
Hm we had discussed this case before landing https://github.com/pytorch/pytorch/pull/107641, but the consensus was that since there are two types of masks 
1.  full boolean attention masks
2. additive floating point attention masks

and 0 / -inf float-only masks fall in the uncomfortable middle of (1) and (2). 

Although disabling the fast path for all floating point masks may regress people who pass in 0 / -inf only float masks, because they fall in the middle and the check might be rather expensive, we should encourage people passing these masks to use a bool mask instead and keep the floating point masks to additive masks

That said, maybe people who use the fast path are largely passing in 0/-inf floating point masks. 

cc @jbschlosser @drisspg for thoughts

**Comment by mikekgfb** [comment]
I’m rather opposed to encouraging the use of bool masks to get to use fast path because bool is not supported by either flash or xformer memory efficient attention?

We give a lot of contradictory guidance, where we tell users to use a particular style in some scenarios and another style in another scenario.

This makes it hard both cognitively, and also to build a cogent end to end application of the types need to change for efficient training vs efficient inference, especially in a network that has other components too, and where it’s not as simple as changing the mask type at invocation
